### DISCUSSION TITLE
Feature: TSWoW

### DISCUSSION THREAD

**Discussion by ihm-tswow** [azerothcore/azerothcore-wotlk#16415] [General]
# TSWoW

_tldr: As some users have misunderstood this issue: TSWoW is **not** working on azerothcore support. This issue is here in case **someone else** wants to port it, and lists requirements and how we can assist such an effort._

We have been asked by @FrancescoBorzi  to provide you with what we know about the possibility of porting the TSWoW modding framework to the AzerothCore emulator. The purpose of this issue is to unify discussions for this endeavor, but you may wish to branch this up into a milestone or project if you do indeed start porting it. This text has been written and signed by me and the @tswow-admin. 

## What is TSWoW?

TSWoW is a modding framework for Wrath of the Lich King that aims to unify content creation and scripting, and is centered around the VSCodium editor and TypeScript programming language. Perhaps our most famous feature is the creation of custom classes with roughly 20 lines of code from the user, offering similar intuitive procedures for creatures, spells, items, quests and gossips, and much more. TSWoW can do a lot of really cool things, but it is a highly experimental and extensive project that is maintained by just two people (myself and @tswow-admin), and we **highly** recommend that anyone interested in porting it **familiarize** themselves with our tool, community and development first, and then seriously consider if the benefit of a port is worth the amount of work it requires.  To learn about our framework, we recommend going through the current TrinityCore instructions on [our wiki](https://tswow.github.io/tswow-wiki/home/).

TSWoW is itself primarily targeting TrinityCore and this a priority that will not change, as we are dedicated to always provide our users with the option of a GPLv3 license without the affero clause. We still welcome other emulators (including affero) to port and use our framework, and are able to offer both support, as time permits, and acommodations in our source to make such ports easier (ifdefs, flags and submoduling where possible to avoid excessive forking, for example).

# Port

TSWoW was originally written with AzerothCore in mind, but this was scrapped early in development and a lot has changed since then. We still believe we have a fairly good idea on what parts are incompatible, which we will list here. Of course, we cannot **ensure** this list is exhaustive, but it should address the vast majority of issues. 

## Database
The first half of TSWoW is an interface for procedurally generating DBC and SQL patches which we call **data scripting**. The primary issue here stems from that our API assumes the TrinityCore database structure. 

- [ ] Database compatibility: The closer your database structure is to TrinityCore, the better. When TSWoW executes data scripts, we both read from and write to the world database and automatically transform this into JavaScript objects that users can manipulate. The more important the table, the more important it is that it's the same since more functionality will be tied into it.
    - If a data script touches a table, every column must match or the script will throw an error even if the column is untouched. You **cannot** get partial compatibility on a per-column basis, as we do not and likely will not track such changes.
    - If a data script does not touch a table, no error should be thrown. You **can** get partial compatibility on a per-table basis.
    - The alternative to adapting your database structure is to fork and modify the TSWoW API. We do not recommend this, since it may render user modules incompatible between emulators, but may be necessary during transition or for more fringe tables. This is trickier than, for example, the Eluna API since it's such a large amount of highly specific database fields that are expected to behave in highly specific ways.

- [ ] Handling incompatibility: It may be possible to allow some limited incompatibility in our API (emulator flags in the API itself), but we do not have a model ready for this. Any suggestions are welcome, but make sure you are first familiar with our models as there are some acommodations we cannot make. 

- [ ] Database generation: We currently generate TrinityCores databases for them, as we need them to be available without starting the worldserver. You need to consider how we should generate yours. Here, we are willing to allow you in our main repository with emulator target flags (i.e. `if(isAzerothCore)` checks) so you don't need to fork just for that. 

## Core
The second half of TSWoW is a scripting interface that uses a [TypeScript->C++ transpiler](https://github.com/tswow/typescript2cxx/tree/tswow) and the TrinityCore script hotswapping system to provide a mix of Eluna-style scripting and core scripts that we call **livescripting**. 

- [x] Script reloader (dll/so loader): Contrary to what some have previously believed, you do **not** need to fully implement TrinityCores script reloading system just to load TSWoW scripts. It is enough to have something that can watch the file system and load/unload dll/so files, which is far easier. However, some early tests have suggested that shared libraries built against your core libraries become up to 100 times larger than for TrinityCore, so this may be something you want to look into even if it isn't a hard requirement. 

- [ ] [TSWoW API](https://github.com/tswow/TrinityCore/tree/tswow/src/server/game/Tswow/Scripting): Instead of exposing TrinityCore to our scripts directly, we already use a wrapper layer to reduce the amount of headers required for users to build scripts, which is good news for you since it's based on Eluna and thus already largely compatible with AzerothCore. We still highly recommend you get in touch with us if working on this issue.
   - [x] Submodule the TSWoW scripting API: Currently the TSWoW API is directly in the TrinityCore source tree.
   - [ ] Fix AzerothCore compatibility: Our API is directly derived from Eluna, but it was hit with regex sledgehammers quite a few times during development since our execution model is completely different, so AzerothCore support and ifdefs need to be restored in a few places. 

- [ ] Core Edits: TSWoW tries to minimize the amount of core edits we do, and they mostly amount to a few fields added to game entities and events that hand over control to our own systems. You can very easily find lines tswow has edited in the TrinityCore source by using git, and will frequently see `// @tswow-begin` comments that state our intentions. A few of these include modifications to enable custom content, such as [unlocking custom classes](https://github.com/tswow/TrinityCore/blob/0c5250cea9a6cf67e35dcf4a232150ef4084e1b6/src/server/shared/SharedDefines.h#L143) and [enabling custom languages](https://github.com/tswow/TrinityCore/blob/0c5250cea9a6cf67e35dcf4a232150ef4084e1b6/src/server/game/Globals/ObjectMgr.cpp#L184).

- [x] Boost: We use boost libraries in a few places. You will have to decide whether to finish your boost implementation first, but we will also accept ifdefs in our source if you deem it necessary. 

# Maintenance

You should consider how to maintain a port once completed. I think it's very unlikely that TSWoW will make any more fundamental changes that would require major refactoring on your end once you're already in, but there can still be some breaking changes mandated by TrinityCores API or other. You would also have to decide how to manage forks, submodules and where to store your own core edits.

We can offer our best efforts to help out when we do update our API as time permits us to. For us to stay co-operative however, we ask one thing from you: please refrain from applying (only) the AGPL to new tswow-specific modules or patches, as this would block us from merging them back. We wouldn't (and couldn't) stop you from doing this, but we also wouldn't help you if you do.

You can contact me or the @tswow-admin either via the AzerothCore or [TSWoW](https://discord.gg/M89n6TZh9x) discord server.

**Comment by ihm-tswow** [comment]
**datascripts**: I am looking at moving SQL definitions from our main repository into our standard library, which would make it far easier to fork them together, since you will likely have to deploy your own standard library.

**Comment by ihm-tswow** [comment]
**livescripts**: The tswow core module is now split up from our TrinityCore fork and can be worked on separately: https://github.com/tswow/tswow-core-module

This repository works similarly to the Eluna main repository (on which it is originally based). However, as previously noted, it has been heavily modified using automated scripts, meaning a lot of `#ifdefs` are broken and need to be fixed up again for azerothcore. My recommendation is that we try to keep these together using ifdefs instead of forks to keep it simple.

**Comment by ihm-tswow** [comment]
Just got this random idea that i'll leave here so I don't forget about it. 

Here's a design suggestion for how we could support multiple database structures in the same spec files (avoiding having to branch it, and possibly having the same stdlib for multiple core targets):

for a table that's unique to a specific core:

```ts
@CoreSpecific("azerothcore")
export class some_azerothcoreRow extends SqlRow<...> {
}
```

for columns that's unique to specific cores:
```ts
export class some_sharedRow extends SqlRow<...> {
        @CoreSpecific("azerothcore")
        get acColumn() { ... }

        @CoreSpecific("trinitycore")
        get tcColumn() { ... }
}
```

It would then be possible to check for illegal row/column access when running datascripts, and to filter read/write queries without noticably affecting performance

**Comment by ihm-tswow** [comment]
> **livescripts**: The tswow core module is now split up from our TrinityCore fork and can be worked on separately: https://github.com/tswow/tswow-core-module
> 
> This repository works similarly to the Eluna main repository (on which it is originally based). However, as previously noted, it has been heavily modified using automated scripts, meaning a lot of `#ifdefs` are broken and need to be fixed up again for azerothcore. My recommendation is that we try to keep these together using ifdefs instead of forks to keep it simple.

~~A small update on this and an update in general I guess. I would strongly discourage any attempts to start porting tswow until our 0.13 release with the massive overhaul of almost every system that it will involve.~~

TSWoW has now been updated to 0.13, and most systems are finalized in terms of api changes.

For this module specifically, it is now located in the tswow root repository instead of its own submodule

**Comment by FrancescoBorzi** [comment]
> would strongly discourage any attempts to start porting tswow until our 0.13 release

thanks @ihm-tswow for letting us know!

**Comment by ihm-tswow** [comment]
Tswow can now start azerothcore (a small step, there's no modding support yet):
https://github.com/tswow/tswow/commit/116b0512a035009d14012ba43589aba4a4369da3

I also wrote a script to compare the databases built for azerothcore and trinitycore. These are all the remaining table/column discrepancies. Not all of these make sense to change (and some tables/columns come from tswow itself), but if you're intereseted in integration you can look over this list if there's _anything_ you'd like to change for compatibility, a few things like the `npc_text` and `spell_proc` stuff seem to just come down to different naming conventions.

_Remember that we don't require full compatibility, but we do need some special handling for tables with missing/very different column layouts. Missing tables should be mostly fine._

[TrinityCore (tswow fork)/AzerothCore table comparison](https://github.com/azerothcore/azerothcore-wotlk/files/8212615/comparison.txt)

**Comment by FrancescoBorzi** [comment]
nice work @ihm-tswow ! 

also about aligning the DB structure, we addressed many of them in the past (see https://github.com/azerothcore/azerothcore-wotlk/issues/1118) but however it would be nice to address the ones that have no reason to be different, such as:

>  a few things like the npc_text and spell_proc stuff seem to just come down to different naming conventions.

if you have a list of those (specific ones), would be nice to create a ticket with them

Keep up the good work 👍

**Comment by ihm-tswow** [comment]
# Big things

AzerothCore can now load both basic [livescripts](https://github.com/tswow/tswow/commit/032ef051c46be60534b8093ddb41c62e579f8b95
) and [datascripts](https://github.com/tswow/tswow/commit/c09eec678ff0968703133be2b0165e1efcc9b916)!

Additionally, i've spent the weekend writing up a larger collection of issues that needs addressing to get azerothcore into a usable state with tswow, which has resulted in the [azerothcore-compatibility](https://github.com/tswow/tswow/milestone/3
) milestone on our tracker.

With this, we're happy to declare we're getting ready to invite any azerothcore developers interested in tswow to [come talk with us](https://discord.gg/M89n6TZh9x) if you want to see a full crossover happen! 

Please keep in mind that getting azerothcore into a **usable state** with tswow is still a long way to go, and that addressing these issues will require varying degrees of existing knowledge with tswow.  Luckily, we already have a few tswow users interested in helping out, so we hope to see some more development on this before long!

**Comment by FrancescoBorzi** [comment]
@ihm-tswow that's great news indeed!
